test/opt/local_single_block_elim.cpp

```diff
@@ -1118,6 +1118,182 @@ TEST_F(LocalSingleBlockLoadStoreElimTest, VariablePointerTest) {
   )";
   SinglePassRunAndMatch<LocalSingleBlockLoadStoreElimPass>(text, false);
 }
+
+TEST_F(LocalSingleBlockLoadStoreElimTest, DebugInfoTest) {
```

Comment:
This is a good test.  We should add another one where a DebugValue with a dref expression is used in place of the DebugDeclare.  We would expect the same resutls.

Reply:
I admit that the OpenCL.DebugInfo.100 spec is vague, but after some research and thought, I do not believe that a DebugValue with a deref of a pointer Value is the same as a DebugDeclare. A DebugValue represents an assignment of the Value and Expression to the **rvalue** of the LocalVariable. A DebugDeclare represents an "equivalencing" of the Expression applied to lvalue of the Variable to the **lvalue** of the LocalVariable. I am fairly certain these are not the same thing.

I am also fairly certain that we will never have cause to have a deref in an Expression in a DebugValue in a shader. The Variable would have to be of type pointer into function scope memory which don't exist in shaders.

Reply:
After some thought, while I am still fairly certain that one cannot use a DebugValue to create an equivalent to a DebugDeclare in a shader, I will grant that one could create a legitimate DebugValue that contains a deref of a local pointer for a shader, However, I think that such a reference would be so highly unlikely that it would not be worth creating the logic to add it to the supported references for this optimization. If such a thing did appear, it would merely disable optimization and not generate incorrect code. So I think the existing logic is reasonable and acceptable at this time. If someone can make an argument for why such a construct would be prevalent enough to justify optimization here, we can reconsider this logic at that time.

Reply:
> A DebugValue represents an assignment of the Value and Expression to the rvalue of the LocalVariable.

I have thought about this a lot as well when trying to figure out the design for scalar replacement.  I understand your point of view.  At first, I was thinking that way as well.  However, I started thinking about how this would actually be implemented by a debugger, and in dwarf.  In DWARF, you don't give the value of a variable.  DWARF say where you go to get the value (register or memory).  The important point to realize is that the DebugValue is not an assignment of a value.  This can be easily confused when rerefs are not used because, in llvm-ir and in spir-v, the ssa value cannot be changed.

The other reason that I believed the llvm.dbg.value could not be viewed as an assignment of a value is the extra work this would put on debuggers, and how it would slow down debugging.  The debugging info is not present in the machine code.  If the debugger was required to capture the value of the expression at every llvm.dbg.value, the debugger would have to stop the normal execution of the program at every position of an llvm.dbg.value to read the value.  This would cause a ridiculous amount of overhead in debugging.  We wouldn't want a spir-v debugger to have to do the same.

Granted this was just guesses, which is why I ask Jaebaek to look at what code actually gets generated by llvm-ir when a debug value is used with a dref expression, and to examine what a real debugger does with it.  We hand wrote llvm-ir that has exactly one llvm.dbg.value where the address of the memory location is the value and there is a deref expression.  The instruction was placed after the first store to the memory location.  There was a second store to the same memory location, but there was not a second llvm.dbg.value.  When running in the debugger, you see the value of the first store when you break after the first store.  However, when you break after the second store, you see the result of the second store.  The debugger did not capture the value of the first store to be used anytime the developer asks for the value of the variable.  The debugger checks the value in the memory location.

Reply:
OK. Thanks for the detailed response. I do now agree that a DebugDeclare can be replaced with a DebugValue with a deref, and I apologize that it took me this long to come to this understanding.

So I guess now the question in my mind is: is this really what we want to do?

Essentially, to do this, we have to keep backing memory for all locals (function scope variables) and keep all Stores to that memory that we normally remove. So we will have to add extra logic to all optimizations that eliminate stores. In one way it will be simpler for the optimizer since it only has to generate one DebugValue at the top of the program, but there will be performance cost from the extra Stores.

This design will also require a whole additional and redundant memory analysis of derefs in DebugValues.

The next question is: do we need to do this? You opined above that DebugValue would be intractable for debuggers as I imagined them. But if you look at https://llvm.org/docs/SourceLevelDebugging.html#object-lifetime-in-optimized-code it appears that llvm.dbg.value is being used in just this manner.

As far as debugger implementation, I imagine the debugger would be able to create a fairly simple map for each local variable, mapping a line number range to a register.

Still, your design will functionally work, albeit with probable performance costs, and I can work with it.

Reply:
I think I can just look at the Value field of the DebugValue and see if its a pointer, so that should be fairly easy.

Reply:
Your questions are very important questions to remove the technical debt and make a reasonable design. I appreciate it. I am conducting more experiments to figure out the relationship between `llvm.dbg.declare` and `llvm.dbg.value` particularly with `DW_OP_LLVM_fragment`. I am also checking `DW_OP_bit_piece`, which is similar to `DW_OP_LLVM_fragment`.

IIRC, we decided to use `DebugValue` with `Deref` to express the declaration because we thought it is the only way to express the declaration in the scalar replacement. In LLVM world, SROA conducts both the scalar replacement and ssa-rewrite in a single pass, so it does not have to consider "splitting a composite type while keeping the information of declaration". On the other hand, spirv-opt scalar replacement must "split a composite type while keeping the information of declaration". Since `DebugDeclare` does not have `Indexes` operand while `DebugValue` does, we had no idea how to use `DebugDeclare` for each member of a composite. See [DebugDeclare](https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.DebugInfo.100.html#DebugDeclare).

The main questions I currently have are
- **Q1**: Does LLVM SROA pass propagate `llvm.dbg.value` for store and phi instructions when we give `llvm.dbg.value` with `DW_OP_deref`?
- **Q2**: Can we use `DW_OP_LLVM_fragment` with `llvm.dbg.declare` for the scalar replacement?
- **Q3**: Is there a way to express a member of a composite using `DebugDeclare` and `Deref` operation?

If **Q2** and **Q3** are true, I think we do not have to use `DebugValue` with `Deref` for the declaration in the scalar replacement.
In addition, if **Q2** and **Q3** are true and **Q1** is false, it would be better to follow what LLVM does i.e., we must not use `DebugValue` with `Deref` as a declaration.

In my brief experiments, **Q1** and **Q2** are true, but I am doublechecking it if I made some mistakes. I am also digging into LLVM source code.
As far as I understand, `DW_OP_LLVM_fragment` is applied to the variable not its value of `llvm.dbg.value` nor its storage of `llvm.dbg.declare`.
I am curious what `DW_OP_bit_piece` means because I guess (I do not have any evidence yet) that it is applied to the value of `llvm.dbg.value` or the storage of `llvm.dbg.declare`.

For **Q3**, I am discussing the spec change with the initial author of the spec.